lla_to_ecef.cpp

```diff
@@ -0,0 +1,390 @@
+ *
+ * Assumes parameters in the following order on each line:
+ *    - Timestamps as seconds since Unix epoch
+ *    - WGS84 latitude [degrees]
+ *    - WGS84 latitude [degrees]
+ *    - WGS84 altitude [kilometers]
+ *
+ * @param filename  file containing data
+ *
+ * @return vector of PositionLLA structs populated from input file
+ *
+ * @throws Runtime error if file cannot be opened
+ */
+std::vector<PositionLLA> read_csv_lla(const std::string& filename) {
+    std::ifstream file(filename);
+    std::vector<PositionLLA> lla_data;
+
+    if (file.is_open()) {
+        std::string line;
+        while (std::getline(file, line)) {
+            std::stringstream ss(line);
+            std::string value;
+            PositionLLA lla_entry;
+
+            int index = 0;
+            while (std::getline(ss, value, ',')) {
+                switch (index) {
+                    case 0: {
+                        std::string span = value;
+                        size_t pos = span.find('.');
+                        lla_entry.t.tv_sec = std::stol(span.substr(0, pos));
+                        lla_entry.t.tv_nsec =
+                            static_cast<int>(std::stod(span.substr(pos)) * 1e9);
+                        break;
+                    }
+                    case 1:
+                        lla_entry.latitude = std::stod(value);
+                        break;
+                    case 2:
+                        lla_entry.longitude = std::stod(value);
+                        break;
+                    case 3:
+                        lla_entry.altitude = std::stod(value) * KM_TO_M;
+                    default:
+                        break;
+                }
+                index++;
+            }
+
+            lla_data.push_back(lla_entry);
+        }
+
+        file.close();
+    } else {
+        throw std::runtime_error(std::format(
+        "Could not open file {}. Exiting", filename));
+    }
+
+    return lla_data;
+}
+
+/**
+ * get_ecef_vel_at_poi
+ * @brief Calculates ECEF velocity at a point of interest.
+ *
+ * Linear interpolation formula is derived by solving for the unkown y-value
+ * from the equation of slopes:
+ *
+ *          slope = (y - y0) / (x - x0) = (y1 - y0) / (x1 - x0)
+ *
+ * @param lla_data          timestamped LLA data (deg, deg, m)
+ * @param point_of_interest timespec for point of interest
+ * @param print_output      flag to print output to stdout
+ *
+ * @return XYZ ECEF velocity at the point of interest
+ */
+std::vector<double> get_ecef_vel_at_poi(
+    const std::vector<PositionLLA>& lla_data,
+    const timespec& point_of_interest,
+    const bool& print_output) {
+
+  // Search data for point of interest
+  std::vector<int> search_results =
+    find_match_or_nearest(lla_data, point_of_interest);
+
+  std::vector<double> velocity_result;
+
+  // If no exact time match for point of interest was found, interpolate
+  if (search_results[1] == INDEX_ERR) {
```

Comment:
nice macro use, also might be a nit, but I prefer keeping return types consistent for methods

so if i understand correctly (plz note I am skimming this at 10pm), you have this method find_match_or_nearest() where the return type could be a status (INDEX_ERR) or an actual value

I would prefer to have a status enum that is the return type of the method, and the user passes in a variable that stores the return result if there is one

that said, could be just a practice that varies from place to place

Reply:
I agree this is probably my clunkiest thing. I can refine it, I was really trying to just get functionality fast and was being lazy. Good idea to use enum.